vulture_os/toolkit/api_parser/extrahop/extrahop.py

```diff
@@ -0,0 +1,204 @@
+
+    HEADERS = {
+        "Accept": "application/json"
+    }
+
+    def __init__(self, data):
+        super().__init__(data)
+
+        self.extrahop_host = "https://" + data["extrahop_host"].split("://")[-1].rstrip("/")
+
+        self.id = data['extrahop_id']
+        self.secret = data['extrahop_secret']
+
+        self.access_token = data.get('extrahop_access_token')
+        self.expire_date = data.get('extrahop_expire_date')
+
+        self.session = None
+
+    def get_access_token(self):
+        if not all(self.access_token, self.expire_date) or self.expire_date < (timezone.now() + timedelta(minutes=5)):
+            try:
+                url = f"{self.extrahop_host}/oauth2/token"
+                auth = HTTPBasicAuth(self.id, self.secret)
+                response = requests.post(url, auth=auth, data={"grant_type": "client_credentials"})
+                response.raise_for_status()
+                response_json = response.json()
+                self.access_token = response_json["access_token"]
+                self.expire_date = (timezone.now() + timedelta(seconds=int(response_json["expires_in"]))).isoformat()
+            except requests.exceptions.HTTPError as err:
+                raise ExtrahopAPIError(f"Unable de renew access token. Error: {err}")
+            except json.DecodeError:
+                raise ExtrahopAPIError("Token renewal response is not a valid JSON")
+            except KeyError as err:
+                raise ExtrahopAPIError(f"Missing key in token renewal response: {err}")
+            else:
+                if self.frontend:
+                    self.frontend.access_token = self.access_token
+                    self.frontend.expire_date = self.expire_date
+                    self.frontend.save()
+
+        return self.access_token
+
+    def _execute_query(self, url, query, timeout=10):
+        """
+        raw request doesn't handle the pagination natively
+        """
+        retry = 1
+        while retry <= 3:
+            if self.session is None:
+                self._connect()
+
+            msg = f"call {url} with query {query} (retry: {retry})"
+            logger.info(f"[{__parser__}]:__execute_query: {msg}", extra={'frontend': str(self.frontend)})
+            try:
+                response = self.session.post(
+                    url,
+                    json=query,
+                    timeout=timeout,
+                    proxies=self.proxies,
+                    verify=self.api_parser_custom_certificate or self.api_parser_verify_ssl
+                )
+
+            except Exception as e:
+                logger.warning(f"[{__parser__}]:__execute_query: Exception: {e}", extra={'frontend': str(self.frontend)})
+
+            else:
+                if response.status_code == 200:
+                    return response.json()
+                elif response.status_code == 204:
+                    return []
+
+                if "The specified access token is invalid." in response.text:
```

Comment:
don't you have a 40x error code on token authentication failures?

Reply:
nop :(